File: docker-learning/04_multi_stage_optimization/Dockerfile.go
# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
# ■ 파일명: Dockerfile.go
# ■ 목적: Go 멀티스테이지 빌드 - scratch 기반 초경량 이미지
# ■ 비교: golang 이미지 ~800MB → scratch ~10MB
# ■ 빌드: docker build -f Dockerfile.go -t go-optimized .
# ■ 날짜: 2026-03-21
# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■

# ============================================================
# 스테이지 1: 빌드 (Build)
# ============================================================
# Go 컴파일러가 있는 이미지에서 빌드
FROM golang:1.23-alpine AS builder

# 빌드에 필요한 도구 설치
# git = go mod가 의존성을 다운로드할 때 필요
# ca-certificates = HTTPS 통신에 필요한 인증서
RUN apk add --no-cache git ca-certificates

WORKDIR /build

# 의존성 파일 먼저 복사 (레이어 캐싱!)
# go.mod, go.sum이 바뀌지 않으면 다시 다운로드 안 해
COPY go.mod go.sum* ./
RUN go mod download

# 소스코드 복사
COPY . .

# Go 바이너리 컴파일
# CGO_ENABLED=0 = C 라이브러리 의존성 없이 순수 Go로 컴파일
#   → scratch 이미지에서 실행 가능! (C 라이브러리가 없으니까)
# GOOS=linux = Linux용으로 컴파일
# -ldflags="-s -w" = 디버그 정보 제거 → 바이너리 크기 줄이기
#   -s = 심볼 테이블 제거
#   -w = DWARF 디버그 정보 제거
# -o /app = 출력 파일 경로
RUN CGO_ENABLED=0 GOOS=linux go build \
    -ldflags="-s -w" \
    -o /app \
    ./cmd/server

# ============================================================
# 스테이지 2: 프로덕션 (scratch)
# ============================================================
# scratch = 완전히 비어있는 이미지! (0 바이트)
# OS도 없고, 셸도 없고, 아무것도 없어
# Go 바이너리 하나만 넣으면 그것만 실행돼
# 장점: 공격 표면이 거의 없어 (보안 최강!)
# 단점: 셸이 없어서 exec으로 디버깅 불가
FROM scratch

# builder에서 인증서 복사 (HTTPS 호출 시 필요)
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/

# 컴파일된 바이너리 하나만 복사!
# Go는 의존성을 모두 바이너리에 포함하기 때문에 이것만 있으면 돼
COPY --from=builder /app /app

# 포트 문서화
EXPOSE 8080

# 바이너리 직접 실행
ENTRYPOINT ["/app"]

# ============================================================
# 결과:
# golang:1.23 이미지: ~800MB
# 최종 이미지 (scratch + 바이너리): ~5~15MB
#
# 80배 이상 줄어들어!
# 그리고 OS가 없으니 취약점도 거의 없어
# ============================================================

File: docker-learning/04_multi_stage_optimization/Dockerfile.java
# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
# ■ 파일명: Dockerfile.java
# ■ 목적: Java 멀티스테이지 빌드 - JDK 빌드 → JRE 런타임
# ■ 비교: JDK 이미지 ~400MB → JRE 이미지 ~200MB
# ■ 빌드: docker build -f Dockerfile.java -t java-optimized .
# ■ 날짜: 2026-03-21
# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■

# ============================================================
# 스테이지 1: 의존성 캐싱
# ============================================================
# Maven/Gradle 의존성을 먼저 다운로드해서 캐싱
# pom.xml이 바뀌지 않으면 다시 다운로드하지 않아
FROM eclipse-temurin:21-jdk-alpine AS deps

WORKDIR /build

# Maven Wrapper 복사 (프로젝트에 포함된 Maven)
COPY mvnw pom.xml ./
COPY .mvn .mvn

# 의존성만 다운로드 (소스코드 컴파일은 안 해)
# -B = batch mode (대화형 입력 없이)
# dependency:go-offline = 모든 의존성을 미리 다운로드
RUN chmod +x mvnw && \
    ./mvnw dependency:go-offline -B

# ============================================================
# 스테이지 2: 빌드 (Build)
# ============================================================
FROM eclipse-temurin:21-jdk-alpine AS builder

WORKDIR /build

# deps 스테이지에서 다운로드한 의존성 복사
COPY --from=deps /root/.m2 /root/.m2

# 전체 소스코드 복사
COPY . .

# 애플리케이션 빌드
# package = 컴파일 + 테스트 + JAR 파일 생성
# -DskipTests = 테스트 건너뛰기 (CI/CD에서 별도로 하니까)
# spring-boot:repackage = 실행 가능한 fat JAR 생성
RUN chmod +x mvnw && \
    ./mvnw package -DskipTests -B

# JAR 파일에서 레이어 추출 (Spring Boot layered JAR)
# Spring Boot 2.3+부터 JAR을 레이어로 분리할 수 있어
# → Docker 캐싱 효율이 올라가 (변경된 레이어만 다시 빌드)
RUN mkdir -p /extracted && \
    java -Djarmode=layertools -jar target/*.jar extract --destination /extracted

# ============================================================
# 스테이지 3: 프로덕션 (JRE만!)
# ============================================================
# JDK = 개발 도구 포함 (컴파일러, 디버거 등) → 크고 불필요
# JRE = 실행만 가능 → 작고 충분해!
FROM eclipse-temurin:21-jre-alpine AS production

# 보안: 일반 사용자 생성
RUN addgroup -S app && adduser -S -G app app

WORKDIR /app

# Spring Boot 레이어를 순서대로 복사
# 아래부터 위로 갈수록 자주 바뀌는 것
# → 자주 바뀌지 않는 것을 먼저 복사 = 캐시 효율 극대화!
COPY --from=builder --chown=app:app /extracted/dependencies/ ./
COPY --from=builder --chown=app:app /extracted/spring-boot-loader/ ./
COPY --from=builder --chown=app:app /extracted/snapshot-dependencies/ ./
COPY --from=builder --chown=app:app /extracted/application/ ./

# 레이어 순서 설명:
# 1. dependencies = 외부 라이브러리 (거의 안 바뀜)
# 2. spring-boot-loader = 부트 로더 (거의 안 바뀜)
# 3. snapshot-dependencies = 개발 중인 라이브러리 (가끔 바뀜)
# 4. application = 우리 코드 (자주 바뀜) → 맨 마지막!

# JVM 최적화 환경변수
# 컨테이너 환경에 맞는 JVM 설정
ENV JAVA_OPTS="-XX:+UseContainerSupport \
    -XX:MaxRAMPercentage=75.0 \
    -XX:InitialRAMPercentage=50.0 \
    -Djava.security.egd=file:/dev/urandom"
# UseContainerSupport = 컨테이너의 메모리 제한을 JVM이 인식
# MaxRAMPercentage = 컨테이너 메모리의 75%까지 사용
# urandom = 빠른 난수 생성 (시작 속도 향상)

# 포트 문서화
EXPOSE 8080

# 헬스체크
HEALTHCHECK --interval=30s --timeout=10s --retries=3 --start-period=40s \
    CMD wget --no-verbose --tries=1 --spider http://localhost:8080/actuator/health || exit 1

# 일반 사용자로 전환
USER app

# Spring Boot 실행
# JarLauncher = Spring Boot의 레이어 구조를 이해하는 런처
ENTRYPOINT ["sh", "-c", "java $JAVA_OPTS org.springframework.boot.loader.launch.JarLauncher"]
